sign/rpmsignverity.c

```diff
@@ -0,0 +1,234 @@
+			       char *keypass, char *cert, uint16_t algo,
+			       uint32_t block_size)
+{
+    struct libfsverity_merkle_tree_params params;
+    struct libfsverity_signature_params sig_params;
+    struct libfsverity_digest *digest = NULL;
+    rpm_loff_t file_size;
+    char *digest_hex, *sig_hex = NULL;
+    uint8_t *sig = NULL;
+    int status;
+
+    if (S_ISLNK(rpmfiFMode(fi)))
+	file_size = 0;
+    else
+	file_size = rpmfiFSize(fi);
+
+    memset(&params, 0, sizeof(struct libfsverity_merkle_tree_params));
+    params.version = 1;
+    params.hash_algorithm = algo;
+    params.block_size = block_size;
+    params.salt_size = 0 /* salt_size */;
+    params.salt = NULL /* salt */;
+    status = libfsverity_compute_digest(fi, file_size, rpmVerityRead,
+					&params, &digest);
+    if (status) {
+	rpmlog(RPMLOG_DEBUG, _("failed to compute digest\n"));
+	goto out;
+    }
+
+    digest_hex = pgpHexStr(digest->digest, digest->digest_size);
+    rpmlog(RPMLOG_DEBUG, _("file(size %li): %s: digest(%i): %s, idx %i\n"),
+	   file_size, rpmfiFN(fi), digest->digest_size, digest_hex,
+	   rpmfiFX(fi));
+
+    free(digest_hex);
+
+    memset(&sig_params, 0, sizeof(struct libfsverity_signature_params));
+    sig_params.keyfile = key;
+    sig_params.certfile = cert;
+    if (libfsverity_sign_digest(digest, &sig_params, &sig, sig_size)) {
+	rpmlog(RPMLOG_DEBUG, _("failed to sign digest\n"));
+	goto out;
+    }
+
+    sig_hex = pgpHexStr(sig, *sig_size);
+ out:
+    free(digest);
+    free(sig);
+    return sig_hex;
+}
+
+rpmRC rpmSignVerity(FD_t fd, Header sigh, Header h, char *key,
+		    char *keypass, char *cert)
+{
+    int rc;
+    FD_t gzdi;
+    rpmfiles files = NULL;
+    rpmfi fi = NULL;
+    rpmfi hfi = rpmfiNew(NULL, h, RPMTAG_BASENAMES, RPMFI_FLAGS_QUERY);
+    rpmts ts = rpmtsCreate();
+    struct rpmtd_s td;
+    off_t offset = Ftell(fd);
+    const char *compr;
+    char *rpmio_flags = NULL;
+    char *sig_hex;
+    char **signatures = NULL;
+    size_t sig_size;
+    uint32_t siglen;
+    int nr_files, idx;
+    uint16_t algo;
+    uint32_t block_size, algo32;
+
+    Fseek(fd, 0, SEEK_SET);
+    rpmtsSetVSFlags(ts, RPMVSF_MASK_NODIGESTS | RPMVSF_MASK_NOSIGNATURES |
+		    RPMVSF_NOHDRCHK);
+
+    rc = rpmReadPackageFile(ts, fd, "fsverity", &h);
+    if (rc != RPMRC_OK) {
+	rpmlog(RPMLOG_DEBUG, _("%s: rpmReadPackageFile returned %i\n"),
+	       __func__, rc);
+	goto out;
+    }
+
+    rpmlog(RPMLOG_DEBUG, _("key: %s\n"), key);
+    rpmlog(RPMLOG_DEBUG, _("cert: %s\n"), cert);
+
+    compr = headerGetString(h, RPMTAG_PAYLOADCOMPRESSOR);
+    rpmio_flags = rstrscat(NULL, "r.", compr ? compr : "gzip", NULL);
+
+    gzdi = Fdopen(fdDup(Fileno(fd)), rpmio_flags);
+    free(rpmio_flags);
+    if (!gzdi)
+	rpmlog(RPMLOG_DEBUG, _("Fdopen() failed\n"));
+
+    files = rpmfilesNew(NULL, h, RPMTAG_BASENAMES, RPMFI_FLAGS_QUERY);
+    fi = rpmfiNewArchiveReader(gzdi, files,
+			       RPMFI_ITER_READ_ARCHIVE_OMIT_HARDLINKS);
```

Comment:
Mmh. Reading through the entire archive unpacking things as we go is expensive and very much out of the ordinary for signing. 

Could you instead use rpm's file hash algorithm for the purpose, ie if rpm's file digests are sha256 then use that for verity too so you don't need to recalculate?

Reply:
I wish I could, but unfortunately I don't believe it is possible. fsverity generates a Merkle tree (basically a tree of digests) and signs the root hash, and we cannot derive the root sha from the file sha.

This is what I mentioned in here: https://github.com/rpm-software-management/rpm/issues/1121#issuecomment-621421288

Reply:
Right, silly me. I plead ignorance and amnesia on what little I know about the Merkle tree stuff... but now that you remind me, it makes me think there's quite a bit of mutual interest here.

There are multiple places in rpm that would benefit from gradually verifiable content, starting with the file digests themselves. If rpm stored the Merkle hashes for the files at build time, I suppose you could then just sign those? And when available, rpm could use those instead of the traditional digests for its verify operation for quicker identification of modified content.

Reply:
> Right, silly me. I plead ignorance and amnesia on what little I know about the Merkle tree stuff... but now that you remind me, it makes me think there's quite a bit of mutual interest here.
> 
> There are multiple places in rpm that would benefit from gradually verifiable content, starting with the file digests themselves. If rpm stored the Merkle hashes for the files at build time, I suppose you could then just sign those? And when available, rpm could use those instead of the traditional digests for its verify operation for quicker identification of modified content.

Yes that should be feasible, the traditional digest can be derived from the Merkle tree. If we added support for rpm to use a Merkle tree internally that would be a mutual benefit, since it allows one to verify a data block in linear time without having to checksum the entire file. Note that the Merkle tree consumes roughly 1/127 of the file size, so there is a cost with it, but it's not crazy. It would be a fair amount of work on top of this, but maybe something to put on the long term goal list?

Reply:
Ok, good to know.
Absolutely not expecting any such work to be part of this effort, just musing about possibilities. It's useful to be at least aware of the possibilities/consequences from the go.